File: src/main/java/uk/co/alt236/apkcompare/comparators/results/comparisons/DateComparison.java
package uk.co.alt236.apkcompare.comparators.results.comparisons;

import uk.co.alt236.apkcompare.comparators.results.Similarity;
import uk.co.alt236.apkcompare.util.date.IsoISO8601DateParser;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.Date;
import java.util.Objects;

public class DateComparison implements Comparison {

    private final String title;
    private final Date value1;
    private final Date value2;
    private final String comparedAttribute;

    public DateComparison(@Nonnull String title,
                          @Nullable String comparedAttribute,
                          @Nullable Date value1,
                          @Nullable Date value2) {
        this.title = title;
        this.comparedAttribute = comparedAttribute;
        this.value1 = value1;
        this.value2 = value2;
    }


    @Override
    public String getTitle() {
        return title;
    }

    @Nullable
    @Override
    public String getValue1AsString() {
        return value1 == null ? null : toIsoDate(value1);
    }

    @Nullable
    @Override
    public String getValue2AsString() {
        return value2 == null ? null : toIsoDate(value2);
    }

    @Nullable
    @Override
    public String getComparedAttribute() {
        return comparedAttribute;
    }

    @Nullable
    public Date getValue1() {
        return value1;
    }

    @Nullable
    public Date getValue2() {
        return value2;
    }

    @Override
    public Similarity getSimilarity() {
        if (Objects.equals(value1, value2)) {
            return Similarity.IDENTICAL;
        } else {
            return Similarity.DIFFERENT;
        }
    }

    private String toIsoDate(final Date date) {
        return IsoISO8601DateParser.toIsoDateString(date);
    }

}

File: app/src/main/java/uk/co/alt236/apkcompare/app/comparators/results/comparisons/DateComparison.kt
package uk.co.alt236.apkcompare.app.comparators.results.comparisons

import uk.co.alt236.apkcompare.app.comparators.results.EnumEvaluators
import uk.co.alt236.apkcompare.app.comparators.results.MissingValue
import uk.co.alt236.apkcompare.app.comparators.results.Similarity
import uk.co.alt236.apkcompare.app.util.date.IsoISO8601DateParser
import java.util.*

class DateComparison(override val title: String,
                     override val comparedAttribute: String?,
                     val value1: Date?,
                     val value2: Date?) : Comparison {

    override val missingValue: MissingValue by lazy { EnumEvaluators.evaluateMissingValue(this) }

    override val value1AsString = if (value1 == null) null else toIsoDate(value1)

    override val value2AsString = if (value2 == null) null else toIsoDate(value2)

    override val similarity: Similarity
        get() = if (value1 == value2) {
            Similarity.IDENTICAL
        } else {
            Similarity.DIFFERENT
        }

    private fun toIsoDate(date: Date): String {
        return IsoISO8601DateParser.toIsoDateString(date)
    }

}